api/dps/index_test.go

```diff
@@ -0,0 +1,456 @@
+
+	if assert.NotNil(t, index) {
+		assert.Equal(t, mock, index.client)
+	}
+}
+
+func TestIndex_First(t *testing.T) {
+	testHeight := uint64(42)
+
+	t.Run("nominal case", func(t *testing.T) {
+		index := Index{
+			client: &apiMock{
+				GetFirstFunc: func(_ context.Context, in *GetFirstRequest, _ ...grpc.CallOption) (*GetFirstResponse, error) {
+					assert.NotNil(t, in)
+
+					return &GetFirstResponse{
+						Height: testHeight,
+					}, nil
+				},
+			},
+		}
+
+		got, err := index.First()
+
+		if assert.NoError(t, err) {
+			assert.Equal(t, testHeight, got)
+		}
+	})
+
+	t.Run("handles indexing failure", func(t *testing.T) {
+		index := Index{
+			client: &apiMock{
+				GetFirstFunc: func(_ context.Context, in *GetFirstRequest, _ ...grpc.CallOption) (*GetFirstResponse, error) {
+					assert.NotNil(t, in)
+
+					return nil, mocks.DummyError
+				},
+			},
+		}
+
+		_, err := index.First()
+		assert.Error(t, err)
+	})
+}
+
+func TestIndex_Last(t *testing.T) {
+	testHeight := uint64(42)
+
+	t.Run("nominal case", func(t *testing.T) {
+		index := Index{
+			client: &apiMock{
+				GetLastFunc: func(_ context.Context, in *GetLastRequest, _ ...grpc.CallOption) (*GetLastResponse, error) {
+					assert.NotNil(t, in)
+
+					return &GetLastResponse{
+						Height: testHeight,
+					}, nil
+				},
+			},
+		}
+
+		got, err := index.Last()
+
+		if assert.NoError(t, err) {
+			assert.Equal(t, testHeight, got)
+		}
+	})
+
+	t.Run("handles indexing failure", func(t *testing.T) {
+		index := Index{
+			client: &apiMock{
+				GetLastFunc: func(_ context.Context, in *GetLastRequest, _ ...grpc.CallOption) (*GetLastResponse, error) {
+					assert.NotNil(t, in)
+
+					return nil, mocks.DummyError
+				},
+			},
+		}
+
+		_, err := index.Last()
+		assert.Error(t, err)
+	})
+
+}
+
+func TestIndex_Header(t *testing.T) {
+	testHeader := &flow.Header{
+		ChainID: dps.FlowTestnet,
+		Height:  42,
+	}
+
+	testHeaderB, err := cbor.Marshal(testHeader)
+	require.NoError(t, err)
+
+	t.Run("nominal case", func(t *testing.T) {
+		index := Index{
+			client: &apiMock{
+				GetHeaderFunc: func(_ context.Context, in *GetHeaderRequest, _ ...grpc.CallOption) (*GetHeaderResponse, error) {
+					assert.Equal(t, testHeader.Height, in.Height)
+
+					return &GetHeaderResponse{
+						Height: testHeader.Height,
+						Data:   testHeaderB,
+					}, nil
+				},
+			},
+		}
+
+		got, err := index.Header(testHeader.Height)
+
+		if assert.NoError(t, err) {
+			assert.Equal(t, testHeader, got)
+		}
+	})
+
+	t.Run("handles indexing failures", func(t *testing.T) {
+		index := Index{
+			client: &apiMock{
+				GetHeaderFunc: func(_ context.Context, in *GetHeaderRequest, _ ...grpc.CallOption) (*GetHeaderResponse, error) {
+					assert.Equal(t, testHeader.Height, in.Height)
+
+					return nil, mocks.DummyError
+				},
+			},
+		}
+
+		_, err := index.Header(testHeader.Height)
+
+		assert.Error(t, err)
+	})
+
+	t.Run("handles marshalling failures", func(t *testing.T) {
```

Comment:
I usually always call it `encoding` and `decoding`, since that's industry standard. The `marshal` / `unmarshal` seems to really just be a Go encoder/decoder interface term to differentiate it with the stream equivalents.